File: 1. 단계별로_풀어보기/20~29/25. 이분 탐색/3. 1654.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const [K, N] = inputArr[0].trim().split(' ').map(Number)
const lineArr = inputArr.slice(1).map(Number)

// 길이의 최대값 구하기
const maxLen = Math.max(...lineArr)

process.stdout.write(`${binarySearch(0, maxLen)}`)


function binarySearch(minLen, maxLen) {
  if (minLen === maxLen) {
    return minLen
  }

  if (minLen + 1 === maxLen) {
    if (getSticks(maxLen) === N) {
      return maxLen
    }
    return minLen
  }

  const half = Math.floor((minLen + maxLen) / 2)

  if (getSticks(half) >= N) {
    return binarySearch(half, maxLen)
  } // ::
  else {
    return binarySearch(minLen, half)
  }
}

function getSticks(div) {
  return lineArr.reduce((prev, val) => prev += Math.floor(val / div), 0)
}

File: 1. 단계별로_풀어보기/20~29/25. 이분 탐색/3. 1654.py
import sys

input = sys.stdin.readline
output = sys.stdout.write

K, N = map(int, input().split())
lineArr = list(int(input()) for i in range(K))

def getSticks(div):
  return sum(line // div for line in lineArr)

def binarySearch(minLen, maxLen):
  if minLen == maxLen:
    return minLen

  if minLen + 1 == maxLen:
    return maxLen if getSticks(maxLen) >= N else minLen

  halfLen = (minLen + maxLen) // 2

  if getSticks(halfLen) >= N:
    return binarySearch(halfLen, maxLen)
  else:
    return binarySearch(minLen, halfLen)

output(f"{binarySearch(0, max(lineArr))}")